### PR TITLE
47 Code Quality PiFac

### LINKED ISSUE #47: Improve code quality for PiFacProof
There are a few code quality items to address in `PiFac` proofs.

- [x] Update fields in proof, secret, and input to be more meaningful where possible (maybe pending #40)
- [x] Update field names to be more descriptive and update field docs to include short names.
- [x] ~update new functions to take cloned parameters by value~ not necessary pending #135 
- [x] Add high-level documentation to `PiFacProof` explaining what statements it proves.
- [x] Try combining multiplications in `prove` and `prove_with_transcript` to reduce re-calculations
- [x] Align type names with other refactored ZKPs, including changing input + secret names to `CommonInput` and `ProverSecret`

### PR DESCRIPTION
**Pull request by hridambasu**
Closes #47

### CODE REVIEW

File: src/zkp/pifac.rs [hunk 1]
```diff
@@ -54,13 +76,15 @@ impl PiFacInput {
     }
 }
 
+/// The prover's secret knowledge: the factors p and q of the modulus N where N
+/// = pq.
```
Comment (marsella) [inline src/zkp/pifac.rs:80]: todo: put math in backticks

**Review by marsella — CHANGES_REQUESTED** [review]
Looking pretty good so far. I think it would be helpful to have a working session to discuss potential names for things together.

File: src/zkp/pifac.rs [hunk 2]
```diff
@@ -6,7 +6,16 @@
 // License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 // of this source tree.
 
-//! Implements the ZKP from Figure 28 of <https://eprint.iacr.org/2021/060.pdf>
+//! Implements a zero-knowledge proof that the modulus N can be factored into
+//! two numbers greater than `2^ℓ` for a parameter ell where `ℓ` is
+//! [`parameters::ELL`](crate::parameters::ELL).
```
Comment (marsella) [inline src/zkp/pifac.rs:11]: todo: `N` should be in backticks.

I think it's confusing to use both ℓ and the word "ell". I would rephrase to:
> ...two numbers greater than `2^ℓ`, where `ℓ` is a fixed parameter defined by
//! [`parameters::ELL`](crate::parameters::ELL).

File: src/zkp/pifac.rs [hunk 3]
```diff
@@ -24,43 +33,62 @@ use std::fmt::Debug;
...
+    /// Commitment to randomness alpha and `x` (`A` in the paper).
+    p_mask_commitment: Commitment,
+    /// Commitment to randomness beta and `y` (`B` in the paper).
+    q_mask_commitment: Commitment,
+    /// Combination of commitment to `Q` using ring Pedersen parameter `r` (`T`
+    /// in the paper).
+    // Commitment to q + p's commitment randomness
+    q_link_commitment: Commitment,
+    /// Randomness for commitment.
     sigma: CommitmentRandomness,
```
Comment (marsella) [inline src/zkp/pifac.rs:53]: suggestion: maybe call this `link_randomness` or something to make it clearer which commitment it's supposed to correspond to.

File: src/zkp/pifac.rs [hunk 4]
```diff
@@ -24,43 +33,62 @@ use std::fmt::Debug;
 use tracing::error;
 use zeroize::ZeroizeOnDrop;
 
+/// Proof that the modulus N can be factored into two numbers greater than 2^ell
+/// for a parameter ell.
 #[derive(Debug, Serialize, Deserialize, Clone)]
 pub(crate) struct PiFacProof {
```
Comment (marsella) [inline src/zkp/pifac.rs:39]: todo: In all of hte documentation here, I would remove references to intermediate variables that aren't exposed in the proof (`x`, `y`, `meu`, `neu`, `alpha`, `beta`, etc). Those won't make sense unless someone is looking at the paper directly. 

When the intermediate var is commitment randomness I think you can just not talk about it. A commitment always has commitment randomness but nobody needs to know about it explicitly except in the inline implementation.

When the intermediate var is a mask I think it's fine to also ignore it in the doc and just talk about the thing that is getting masked without naming the mask itself.

File: src/zkp/pifac.rs [hunk 5]
```diff
@@ -24,43 +33,62 @@ use std::fmt::Debug;
...
-pub(crate) struct PiFacInput {
+pub(crate) struct CommonInput {
     setup_params: VerifiedRingPedersen,
-    N0: BigNumber,
+    modulus: BigNumber,
 }
 
-impl PiFacInput {
+impl CommonInput {
+    /// Generate public input for proving and verifying [`PiFacProof`] about N.
```
Comment (marsella) [inline src/zkp/pifac.rs:74]: nit: `N` in backticks

File: src/zkp/pifac.rs [hunk 6]
```diff
@@ -156,64 +196,82 @@ impl Proof for PiFacProof {
...
         };
-        if !eq_check_2 {
+        if !masked_q_commitment_is_valid {
             error!("eq_check_2 failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_3 = {
+        let modulus_links_provided_factors = {
             let R = input
```
Comment (marsella) [inline src/zkp/pifac.rs:245]: todo: rename this something Rust-canonical.

**Review by marsella — CHANGES_REQUESTED** [review]
I think all of these comments are about documentation :)

File: src/zkp/pifac.rs (hunk 4 above)
Comment (hridambasu) [reply src/zkp/pifac.rs:39]: @marsella  But then explaining more things is better no? In some sense, people who understands the paper and looks at the code has better visibility since the var names used in the code and the paper are different and if I completely remove all references to alpha, x, etc. then it would be hard for people to understand and makes the documentation incomplete in my opinion. Let me know your thoughts. If you think it confuses people more, then I would completely get rid of them.

Comment (marsella) [reply src/zkp/pifac.rs:39]: This section should aim to describe the overall structure of the proof, via good names, good types, and succinct docs:
- what are we sending?
- what's the type of each field?
- what's the role of each field in the proof?

I agree that we should try to reference all the variables generated as part of the proof, but I think the details about, say, names of commitment randomness belong inline in the proof -- where we actually create them. Right now many of the names in `prove()` do map directly to the paper, so auditors can tell what's what. There are a few that have gotten more descriptive names (like `p_mask` instead of `alpha`); I think it would be reasonable to add docs at the place where we create those to map to the paper name. 

Does that sound reasonable to you? The goal is that someone looking at this `PiFacProof` type who's familiar with how sigma proofs work will be able to scan it quickly and understand the general shape without getting bogged down in implementation details. If they want those details, they can look at the `prove` and `verify` methods.

File: src/zkp/pifac.rs [hunk 7]
```diff
@@ -156,64 +201,78 @@ impl Proof for PiFacProof {
...
         }
 
-        let sqrt_N0 = sqrt(&input.N0);
-        // 2^{ELL + EPSILON}
-        let two_ell_eps = BigNumber::one() << (ELL + EPSILON);
-        // 2^{ELL + EPSILON} * sqrt(N_0)
-        let z_bound = &sqrt_N0 * &two_ell_eps;
-        if self.z1 < -z_bound.clone() || self.z1 > z_bound {
+        if crate::utils::within_bound_by_size(&self.p_masked, ELL + EPSILON) {
             error!("self.z1 > z_bound check failed");
```
Comment (marsella) [inline src/zkp/pifac.rs:272]: todo: Make these logs more useful. If this fails it means that `p` is out of range. Ditto below for `z2`/`q`.

File: src/zkp/pifac.rs [hunk 8]
```diff
@@ -156,64 +196,82 @@ impl Proof for PiFacProof {
...
         let two_ell_eps = BigNumber::one() << (ELL + EPSILON);
         // 2^{ELL + EPSILON} * sqrt(N_0)
         let z_bound = &sqrt_N0 * &two_ell_eps;
-        if self.z1 < -z_bound.clone() || self.z1 > z_bound {
+        if self.p_masked < -z_bound.clone() || self.p_masked > z_bound {
             error!("self.z1 > z_bound check failed");
             return Err(InternalError::ProtocolError);
         }
-        if self.z2 < -z_bound.clone() || self.z2 > z_bound {
+        if self.q_masked < -z_bound.clone() || self.q_masked > z_bound {
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/pifac.rs]: Oh, now that I look at that function more closely, it checks that the parameter is in the range `± 2^n`. The check we need in this proof doing here is that the parameter is in the range `±√modulus · 2 ^ (ℓ+ε)`. So that function actually doesn't work as-is.

Please either write a separate utility function that tests ranges of the form we need here or revert back to doing it inline here.

File: src/zkp/pifac.rs (hunk 4 above)
Comment (marsella) [reply src/zkp/pifac.rs:39]: Oh, I was imagining a bit more of a middle ground than this. E.g. for the `*_mask_commitment`s I would say
```
/// Commitment to a mask for p (`A` in the paper)
```

and for the `w1/w2`s, something like:
```
/// Masked commitment randomness used to form `p_commitment`.
```

File: src/zkp/pifac.rs (hunk 2 above)
Comment (marsella) [reply src/zkp/pifac.rs]: I still see a `...` and no link to `parameters::ELL`.

File: src/zkp/pifac.rs [hunk 9]
```diff
@@ -24,43 +33,66 @@ use std::fmt::Debug;
...
+    /// Commitment to the factor `q` (`Q` in the paper).
+    q_commitment: Commitment,
+    /// `A` in the paper.
+    p_mask_commitment: Commitment,
+    /// `B` in the paper.
+    q_mask_commitment: Commitment,
+    /// Commitment linking `q` to the commitment randomness used in
+    /// `p_commitment`.
+    q_link_commitment: Commitment,
+    /// Randomness for commitment.
```
Comment (marsella) [inline src/zkp/pifac.rs:51]: todo: which commitment?

File: src/zkp/pifac.rs (hunk 4 above)
Comment (marsella) [reply src/zkp/pifac.rs:39]: Please update the docs for `{p, q}_mask_commitment` and `masked_{p, q}_commitment_randomness` per comment above.

File: src/zkp/pifac.rs [hunk 10]
```diff
@@ -90,58 +122,71 @@ impl Proof for PiFacProof {
...
-            .setup_params
-            .scheme()
-            .commit(&beta, ELL + EPSILON, rng);
-        let (T, r) = input.setup_params.scheme().commit_with_commitment(
-            &Q,
-            &alpha,
+        let sqrt_N0 = &sqrt(&input.modulus);
+
+        let p_mask = random_plusminus_scaled(rng, ELL + EPSILON, sqrt_N0);
+        let q_mask = random_plusminus_scaled(rng, ELL + EPSILON, sqrt_N0);
```
Comment (marsella) [inline src/zkp/pifac.rs:128]: suggestion: you can add an inline comment that these are `alpha` and `beta` in the paper for clarity.

File: src/zkp/pifac.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/pifac.rs]: The change I recommended makes this check incorrect, this needs to be changed.

**Review by marsella — COMMENTED** [review]
There are still some unresolved comments from previous reviews; they need to be addressed before this can be approved.

File: src/zkp/pifac.rs [hunk 11]
```diff
@@ -156,68 +203,86 @@ impl Proof for PiFacProof {
...
+        if self.p_masked < -z_bound.clone() || self.p_masked > z_bound {
+            error!("p is out of range!");
             return Err(InternalError::ProtocolError);
         }
-        if self.z2 < -z_bound.clone() || self.z2 > z_bound {
-            error!("self.z2 > z_bound check failed");
+        if self.q_masked < -z_bound.clone() || self.q_masked > z_bound {
+            error!("q is out of range!");
             return Err(InternalError::ProtocolError);
         }
```
Comment (marsella) [inline src/zkp/pifac.rs:285]: optional nit: You could ditch the clone by negating a reference to the bound, like
```rust
if self.p_masked < -&z_bound || self.p_masked > z_bound {
```
It still looks a big awkward but it avoids the clone at least. Not a big deal though.

**Review by marsella — APPROVED** [review]
looks good, thank you!
